### PR TITLE
Introduce new top level parameter manage_cron_hourly which defaults …

### LINKED ISSUE #136: Unable to disable cron for hourly rotations
The default crontab on a Ubuntu machine runs hourly/daily/weekly/monthly jobs at offsets off of the hour which is highly undesirable for log rotations which should be done on the hour.

For daily jobs we have told the logrotate module to not manage cron, and then used puppet/cron to create an appropriate job that runs at 00:00 each day.

But there is no way to tell logrotate module to not manage cron for hourly rotations.

Is there any chance this module can be transitioned to use puppet/cron which supports creating cron jobs where you can specify the exact hour, minute, day, week, etc.
https://forge.puppet.com/puppet/cron

In theory then logrotate::cron could simply call puppet/cron

The issue I see is the lack of FreeBSD support in puppet/cron and if that is still important to this module then I suggest we provide a flag manage_cron_hourly so hourly jobs be set to be unmanaged.

### PR DESCRIPTION
**Pull request by cliff-wakefield**
#### Pull Request (PR) description
Introduced new top level parameter manage_cron_hourly which defaults to true and controls whether the module manages the cron hourly job

#### This Pull Request (PR) fixes the following issues

Fixes #136

### CODE REVIEW

**Comment by cliff-wakefield** [comment]
As usual I have no idea why the checks fail, the tests and error messages are not helpful.

**Comment by bastelfreak** [comment]
Hi @cliff-wakefield. This bug was caused by a bug in beaker-puppet. It got fixed in  https://github.com/puppetlabs/beaker-puppet/pull/68. I restarted the tests.

**Comment by cliff-wakefield** [comment]
Thanks for the update @bastelfreak

**Comment by cliff-wakefield** [comment]
Any chance we can get this merged now?

File: manifests/init.pp [hunk 1]
```diff
@@ -3,6 +3,7 @@
   String $ensure                         = present,
   Boolean $hieramerge                    = false,
   Boolean $manage_cron_daily             = true,
+  Boolean $manage_cron_hourly            = true,
```
Comment (bastelfreak) [inline manifests/init.pp]: This would be a breaking change, because a before unmanaged file would now be managed and might get purged. In my opintion this should default to ```false```

File: manifests/hourly.pp [hunk 2]
```diff
@@ -16,19 +16,27 @@
...
   file { "${logrotate::rules_configdir}/hourly":
     ensure => $dir_ensure,
     owner  => 'root',
     group  => 'root',
     mode   => $logrotate::rules_configdir_mode,
   }
-  -> logrotate::cron { 'hourly':
-      ensure => $ensure,
-    }
+  logrotate::cron { 'hourly':
```
Comment (bastelfreak) [inline manifests/hourly.pp]: I do not quite get the logic here. The file is managed, not matter which value  ```$manage_cron_hourly``` has. But the file should only be managed if the variable is true?

File: manifests/hourly.pp [hunk 3]
```diff
@@ -16,19 +16,27 @@
   Enum['present','absent'] $ensure = 'present'
 ) {
 
+  $manage_cron_hourly = $::logrotate::manage_cron_hourly
```
Comment (bastelfreak) [inline manifests/hourly.pp]: please do not add the leading ```::```, those are only needed for puppet 3 and are now legacy.

File: manifests/config.pp [hunk 4]
```diff
@@ -15,8 +15,13 @@
     mode    => $logrotate::rules_configdir_mode,
   }
 
-  if $manage_cron_daily {
-    logrotate::cron { 'daily': }
+  $cron_ensure = $manage_cron_daily ? {
```
Comment (bastelfreak) [inline manifests/config.pp]: why was this part changed?

**Comment by bastelfreak** [comment]
@cliff-wakefield I added some inline comments. Please also add some unit or acceptance tests.

File: manifests/hourly.pp (hunk 2 above)
Comment (cliff-wakefield) [reply manifests/hourly.pp]: If manage_cron_hourly is not true then absent must be passed down to logrotate::cron to ensure the cron entry is absent from /etc/cron.hourly.

The whole point of this PR is to allow rotate_every => 'hour' to be used but not have the module manage the associated cron entry for hourly rotations.

Allowing a user to define their own schedule for hourly rotations using puppet::cron and hence dictate when in the hour the rotation occurs, such as on the hour not 17 minutes past the hour which I think is the Ubuntu default.

I would much rather switch this whole module to use puppet::cron and then allow a variable such as logrotate::cron_hourly_minute be used to control when an hourly cron job is run.

But I believe that the adoption of puppet::cron is a larger change than just this PR.

File: manifests/hourly.pp (hunk 3 above)
Comment (cliff-wakefield) [reply manifests/hourly.pp]: Can you point me at the documentation that states these are legacy as I use them extensively throughout my profile and other component modules?

Comment (bastelfreak) [reply manifests/hourly.pp]: We have the following in our [review guidelines](https://voxpupuli.org/docs/#reviewing-a-module-pr):
> * In the majority of cases, variables shouldn’t be accessed via topscope: $::modulename::$param. Instead do: $modulename::$param

This is also somehwere documented within the official docs, but I do not have the link by hand.

File: manifests/config.pp (hunk 4 above)
Comment (cliff-wakefield) [reply manifests/config.pp]: For the same reason I explained below. 

When someone elects to not let the module manage the daily cron entry the entry should be removed from /etc/cron.daily by passing the appropriate ensure value through to logrotate::cron. 

With the code the way it was if someone originally ran with manage_cron_daily = true and then moved to manage_cron_daily = false the /etc/cron.daily entry is not cleaned up.

File: manifests/init.pp (hunk 1 above)
Comment (cliff-wakefield) [reply manifests/init.pp]: Happy to change the default to false

Comment (cliff-wakefield) [reply manifests/init.pp]: That being said in the current code logic if a rule is put in place for hourly then it includes ::logrotate::hourly (note the ::) which automatically added the /etc/cron.hourly entry.

So it automagically managed cron hourly but again if you changed your rule away from hourly it never cleaned up the /etc/cron.hourly entry.

Unless I misundertand the -> operator used.

Comment (cliff-wakefield) [reply manifests/init.pp]: If the default is left to true yes it creates a /etc/cron.hourly entry when there may not be any hourly rules, but logrotate will run without erroring it just won't process anything.

File: manifests/hourly.pp (hunk 3 above)
Comment (cliff-wakefield) [reply manifests/hourly.pp]: Thanks

**Comment by acurus-puppetmaster** [comment]
Any response to my comments @bastelfreak ?

**Comment by cliff-wakefield** [comment]
> Any response to my comments @bastelfreak ?

Sorry my comments...

**Comment by acurus-puppetmaster** [comment]
Can someone from @voxpupuli/collaborators please pick this PR up it has been hanging for a little while now.
